torchtext/csrc/vectors.cpp

```diff
@@ -60,16 +81,195 @@ struct Vectors : torch::CustomClassHolder {
+      // views.
+      vectors_ = at::cat({vectors_, vector.unsqueeze(0)});
     }
   }
 
   int64_t __len__() { return stovec_.size(); }
 };
 
+inline int64_t divup(int64_t x, int64_t y) { return (x + y - 1) / y; }
+
+std::tuple<int64_t, int64_t, int64_t>
+_infer_shape(const std::string &file_path, const int64_t delimiter_ascii) {
+
+  int64_t num_header_lines = 0, num_lines = 0, vector_dim = -1;
+  std::vector<std::string> vec_str;
+  std::string line, word;
+
+  std::ifstream fin;
+  fin.open(file_path, std::ios::in);
+
+  while (std::getline(fin, line)) {
+    vec_str.clear();
+    if (vector_dim == -1) {
+      std::istringstream s(line);
+
+      // get rid of the token
+      std::getline(s, word, static_cast<char>(delimiter_ascii));
+      // we assume entries for vector are always seperated by ' '
+      while (std::getline(s, word, ' ')) {
+        vec_str.push_back(word);
+      }
+
+      // assuming word, [vector] format
+      // the header present in some(w2v) formats contains two elements
+      if (vec_str.size() <= 2) {
+        num_header_lines++;
+      } else if (vec_str.size() > 2) {
+        vector_dim = vec_str.size();
+        num_lines++; // first element read
+      }
+    } else {
+      num_lines++;
+    }
+  }
+  return std::make_tuple(num_lines, num_header_lines, vector_dim);
+}
+
+void _infer_offsets(const std::string &file_path, int64_t num_lines,
+                    int64_t chunk_size, int64_t num_header_lines,
+                    std::vector<size_t> *offsets) {
+
+  std::ifstream fin;
+  fin.open(file_path, std::ios::in);
+
+  while (num_header_lines > 0) {
+    fin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+    num_header_lines--;
+  }
+  (*offsets).push_back(fin.tellg());
+  size_t offset = 0;
+  while (fin.ignore(std::numeric_limits<std::streamsize>::max(), '\n')) {
+    offset++;
+    if (offset % chunk_size == 0) {
+      (*offsets).push_back(fin.tellg());
+    }
+  }
+}
+
+void parse_chunk(const std::string &file_path, size_t offset,
+                 const int64_t start_line, const int64_t end_line,
+                 const int64_t vector_dim, const int64_t delimiter_ascii,
+                 std::shared_ptr<StringList> tokens, float *data_ptr) {
+  std::ifstream fin;
+  fin.open(file_path, std::ios::in);
+  fin.seekg(offset);
+
+  int converter_flags = double_conversion::StringToDoubleConverter::NO_FLAGS;
+  double_conversion::StringToDoubleConverter converter(
+      converter_flags, 0.0f, double_conversion::Single::NaN(), NULL, NULL);
+
+  for (int64_t i = start_line; i < end_line; i++) {
+    std::string token;
+    // read the token
+    std::getline(fin, token, static_cast<char>(delimiter_ascii));
+    tokens->push_back(token);
+
+    std::string vec_val;
+    // read the vector
+    for (int64_t j = 0; j < vector_dim; j++) {
+      fin >> vec_val;
+      const char *tmp_str = vec_val.c_str();
+      int processed_characters_count;
```

Comment:
We could TORCH_CHECK that the number of processed characters matches the length of the string.

Reply:
Will do this in the follow-up PR that adds the double conversion library!